app/main/views/frameworks.py

```diff
@@ -1268,6 +1268,15 @@ def sign_framework_agreement(framework_slug):
     framework_pdf_metadata = {
         'g-cloud-12': {'file_size': '962KB', 'page_count': 65}
     }
+    _drafts, complete_drafts = get_drafts(data_api_client, framework_slug)
+    lots = [
+        dict(lot,
+             complete_count=count_drafts_by_lot(complete_drafts, lot['slug']))
+        for lot in framework['lots']]
```

Comment:
Looking at these (fairly ancient) helper functions, I think we can now do this in a much more efficient way, without having to iterate through 30,000+ draft services (or indeed, any draft services at all!).

Something like:

```
complete_drafts = data_api_client.find_draft_services_by_framework(framework_slug, supplier_id=supplier_id, status="submitted", lot=lot)
on_lot = complete_drafts['meta']['total'] > 0
```

The snag is that we currently we can't filter this API endpoint by `lot`. But this is really easy to add! And would be really useful elsewhere. I'm happy to rustle up that PR.

Reply:
https://github.com/alphagov/digitalmarketplace-api/pull/1049

Reply:
Thanks for implementing this @katstevens. I updated the view to use it in 7f54158.